### PR TITLE
EPMDJ-6051: Add auto reconnecting for socket

### PR DESCRIPTION
**Pull request by IliaSolovev**
Issue: EPMDJ-6051

### CODE REVIEW

File: src/common/connection/drill-socket.ts [hunk 1]
```diff
@@ -43,11 +45,19 @@ export class DrillSocket {
 
   public connection$: Observable<DrillResponse>;
 
+  public autoSubscribe$: Subject<'CLOSE' | 'OPEN'>;
```
Comment (RomanAchilov) [inline src/common/connection/drill-socket.ts:48]: reconnection

**Review by RomanAchilov — APPROVED** [review]

File: src/common/connection/drill-socket.ts [hunk 2]
```diff
@@ -57,40 +67,28 @@ export class DrillSocket {
...
-          callback(responseMessage || null);
-        }
-      },
-    );
+    let subscription = this.connection$.subscribe({ next: subscribe(topic, callback, message) });
+
+    const autoSubscription = this.reconnection$.subscribe((type) => {
+      if (type === 'CLOSE') {
+        subscription.unsubscribe();
+        this.send(topic, 'UNSUBSCRIBE', message);
```
Comment (RomanDavlyatshin) [inline src/common/connection/drill-socket.ts:79]: I believe that does not work as you might expect:
1. The connection is already closed, why `this.send` should work?
2.  I've noticed that these messages **are sent**, but only once the connection is reestablished. This is peculiar. Please, look into that (see first messages in the new socket after reconnect).
3. Sending the 'UNSUBSCRIBE'  message might be pointless:
  - if the backend automatically throws away subscriptions for a disconnected socket (which it should);
  - if the backend is restarted "from scratch" (thus, no previous agents exist).

File: src/common/connection/drill-socket.ts [hunk 3]
```diff
@@ -111,3 +109,32 @@ export class DrillSocket {
     });
   }
 }
+
+const subscribe = (topic: string, callback: (arg: any) => void, message?: SubscriptionMessage | Record<string, unknown>) => ({
```
Comment (RomanDavlyatshin) [inline src/common/connection/drill-socket.ts:113]: I think `subscribe` is a misleading name. It's really just a handler. Can you think of a better name?

File: src/common/connection/drill-socket.ts [hunk 4]
```diff
@@ -43,11 +45,19 @@ export class DrillSocket {
...
       url,
       closeObserver: {
-        next: () => this.onCloseEvent(),
+        next: () => {
+          this.onCloseEvent();
+          this.reconnection$.next('CLOSE');
+        },
+      },
+      openObserver: {
+        next: () => this.reconnection$.next('OPEN'),
```
Comment (RomanDavlyatshin) [inline src/common/connection/drill-socket.ts:60]: We probably should emit `onOpenEvent`, to hide the "The backend is unavailable"  alert automatically, once the connection is reestablished.

**Review by RomanDavlyatshin — CHANGES_REQUESTED** [review]

**Pull request merged**